src/main/java/com/amazon/ion/bytecode/bin11/bytearray/PrimitiveDecoder.kt

```diff
@@ -17,86 +17,75 @@ internal object PrimitiveDecoder {
+        // TODO: ion-java#1114
+        if (source.size < start + 3) throw IonException("Incomplete data: start=$start, length=3, limit=${source.size}")
+        return read3IntBytes(source[start], source, start).shr(8)
     }
 
     @JvmStatic
-    fun ByteArray.readFixedInt8AsShort(start: Int): Short {
+    fun readFixedInt32(source: ByteArray, start: Int): Int {
         // TODO: ion-java#1114
-        if (this.size < start + 1) throw IonException("Incomplete data: start=$start, length=${1}, limit=${this.size}")
-        return this[start].toShort()
+        if (source.size < start + 4) throw IonException("Incomplete data: start=$start, length=4, limit=${source.size}")
+        return read4IntBytes(source[start], source, start)
     }
 
     @JvmStatic
-    fun ByteArray.readFixedInt16AsShort(start: Int): Short {
+    fun readFixedInt64(source: ByteArray, start: Int): Long {
         // TODO: ion-java#1114
-        if (this.size < start + 2) throw IonException("Incomplete data: start=$start, length=${2}, limit=${this.size}")
-        return this.getShort(start)
+        if (source.size < start + 4) throw IonException("Incomplete data: start=$start, length=4, limit=${source.size}")
+        return read8IntBytes(source[start], source, start)
     }
 
     @JvmStatic
-    fun ByteArray.readFixedInt24AsInt(start: Int): Int {
+    fun readFixedIntAsLong(source: ByteArray, start: Int, length: Int): Long {
         // TODO: ion-java#1114
-        if (this.size < start + 3) throw IonException("Incomplete data: start=$start, length=${3}, limit=${this.size}")
-        return this.getInt24(start)
+        if (source.size < start + length) throw IonException("Incomplete data: start=$start, length=$length, limit=${source.size}")
+        val firstByte = source[start]
+        return when (length) {
+            1 -> firstByte.toLong()
+            2 -> read2IntBytes(firstByte, source, start).shr(16).toLong()
+            3 -> read3IntBytes(firstByte, source, start).shr(8).toLong()
+            4 -> read4IntBytes(firstByte, source, start).toLong()
+            5 -> read5IntBytes(firstByte, source, start).shr(24)
+            6 -> read6IntBytes(firstByte, source, start).shr(16)
+            7 -> read7IntBytes(firstByte, source, start).shr(8)
+            8 -> read8IntBytes(firstByte, source, start)
+            // TODO: Technically, it's possible that the FixedInt is over-padded with 0-bytes, but we can deal with that later.
+            else -> throw IonException("FixedInt with length $length is too large to fit in a Long")
+        }
     }
 
     @JvmStatic
-    fun ByteArray.readFixedInt32AsInt(start: Int): Int {
-        // TODO: ion-java#1114
-        if (this.size < start + 4) throw IonException("Incomplete data: start=$start, length=${4}, limit=${this.size}")
-        return this.getInt(start)
+    fun readFixedUInt16(source: ByteArray, position: Int): UShort {
+        return read2IntBytes(source[position], source, position).shr(16).toUShort()
```

Comment:
These versions should also have the check right?

```suggestion
    fun readFixedUInt16(source: ByteArray, position: Int): UShort {
        // TODO: ion-java#1114
        if (source.size < start + 2) throw IonException("Incomplete data: start=$start, length=2, limit=${source.size}")
        return read2IntBytes(source[position], source, position).shr(16).toUShort()
```

Reply:
Sure, I'll add the checks in.
It also occurred to me that in some cases, we don't need to check that there's enough data remaining in the byte array because that's already been proved elsewhere. I'll update #1114 to mention that.